### PR TITLE
Authcall

### PR DESCRIPTION
**Pull request by jflo**
- adds AUTH and AUTHCALL opcodes to Prague evms.
- adds a mutable authorizedBy:Address field to the MessageFrame
- adds AUTH opcode costing to gas calculators

### CODE REVIEW

File: evm/src/main/java/org/hyperledger/besu/evm/MainnetEVMs.java [hunk 1]
```diff
@@ -948,6 +950,8 @@ public static void registerPragueOperations(
       final GasCalculator gasCalculator,
       final BigInteger chainID) {
     registerCancunOperations(registry, gasCalculator, chainID);
+    registry.put(new AuthOperation(gasCalculator));
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/MainnetEVMs.java:953]: Ad a comment referring to the EIP these are for (like in registerCancunOperations)

File: evm/src/main/java/org/hyperledger/besu/evm/operation/AuthCallOperation.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  @Override
+  public OperationResult execute(final MessageFrame frame, final EVM evm) {
+    if (frame.isStatic() && !value(frame).isZero()) {
+      return new OperationResult(cost(frame, true), ExceptionalHaltReason.ILLEGAL_STATE_CHANGE);
+    } else if (frame.getAuthorizedBy() != null) {
+      return super.execute(frame, evm);
+    } else {
+      return new OperationResult(cost(frame, true), null);
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/AuthCallOperation.java:97]: Failures should push `frame.pushStackItem(LEGACY_FAILURE_STACK_ITEM);` (i.e. zero) to the stack prior to returning an `OperationResult`

File: evm/src/test/java/org/hyperledger/besu/evm/operations/AuthOperationsTest.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    when(frame.readMemory(65, 32)).thenReturn(contractCommitment);
+
+    AuthOperation authOperation = new AuthOperation(new PragueGasCalculator());
+    authOperation.execute(frame, fakeEVM);
+    verify(frame).setAuthorizedBy(authingAddress);
+    verify(frame).pushStackItem(UInt256.ONE);
+  }
+
+  @Test
+  public void authCallHappyPath() {
```
Comment (shemnon) [inline evm/src/test/java/org/hyperledger/besu/evm/operations/AuthOperationsTest.java:99]: * Needs failure case
* Both need to assert the top stack value is 1 (success) or 0 (failure)

**Review by shemnon — CHANGES_REQUESTED** [review]

File: evm/src/main/java/org/hyperledger/besu/evm/operation/AuthCallOperation.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+/** Introduced via EIP-3074 to call another contract with a different authorization context. */
+public class AuthCallOperation extends AbstractCallOperation {
+
+  /**
+   * Instantiates a new AuthCallOperation.
+   *
+   * @param gasCalculator a Prague or later gas calculator
+   */
+  public AuthCallOperation(final GasCalculator gasCalculator) {
+    super(0xF7, "AUTHCALL", 7, 1, gasCalculator);
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/AuthCallOperation.java:38]: FYI this  and AUTH conflicts with an EOF allocated opcode.  Fun times ahead.

**Review by shemnon — APPROVED** [review]
Have any official reference test vectors been published?

**Comment by jflo** [comment]
> Have any official reference test vectors been published?

Nope. Am keeping an eye out for them, and will be running them asap.

**Pull request merged**